### ISSUE TITLE
Issues Since Upgrade From v40 to v42

### ISSUE THREAD

**Issue by Danny-James** [mapbox/mapbox-navigation-android#2226]
Since updating from `Version 40` to `Version 42` I've noticed a couple of issue; 

**Issue 1:**
Re-centre button no longer appears, when the map is moved I am unable to get the camera to track the navigation puck.

**Issue 2:** 
Since updating to `Version 42` the Navigation Puck seems to no longer be snapping to the route, see images below;

This Image show's the navigation puck off the screen, with no Re-Centre button available to follow the puck:
![20191011_084836](https://user-images.githubusercontent.com/32104383/67023102-2aad4700-f0fa-11e9-8bb5-bf62fd7cb2c1.jpg)

The below images show the Navigation puck slightly off route, when looking at Google Maps the location seems fine, it just seems to have an issue staying on the route line, which was not an issue with `Version 40`:
![Snapchat-1445326784](https://user-images.githubusercontent.com/32104383/67023268-73650000-f0fa-11e9-949b-3dc4f20c9e5d.jpg)
![20191011_092440](https://user-images.githubusercontent.com/32104383/67023276-75c75a00-f0fa-11e9-96ac-1df92ceeb461.jpg)

As you can see from the Video below; I am driving along the motoway but the puck seems to be doing it's own thing and seems like it's gone off route.
https://streamable.com/i4k4l

**Comment by Danny-James** [comment]
Anybody able to investigate this issue?

**Comment by Danny-James** [comment]
I've just attempted to update to `v0.42.3` with simulate route enabled, if i move the map the reroute button is not visible and the camera is not tracking the navigation puck.

**Comment by Danny-James** [comment]
@danesfeder @Guardiola31337  any chance you could have someone look into this?

**Comment by Guardiola31337** [comment]
Hey @Danny-James 👋 thanks for reaching out and report your issue.

Could you add some more information for reproducing the behavior you're describing or any other specifics around your setup? That would be really helpful because without having additional information, I'm unable to reproduce - I've just tested [test app](https://github.com/mapbox/mapbox-navigation-android/tree/master/app) examples and they're working 👌 Are you able to reproduce that behavior in any of the examples of the [test app](https://github.com/mapbox/mapbox-navigation-android/tree/master/app)? Are you tweaking anything? If there is any sample or pseudo code you can provide, that would be really helpful and it'll give us better insight into your implementation and what could be causing your problem.

Thanks again!

**Comment by Danny-James** [comment]
So after a bit of investigation; I've managed to come up with a workaround, however i'm still unsure why this has changed since `v0.40.0`

```java
        navigationView.startNavigation(navigationViewOptions);
...
        // Reset navigationView Camera Position.
        navigationView.resetCameraPosition();
...
```

By calling `navigationView.resetCameraPosition();` after the `startNavigation` the camera and recenter button function as they should, and allows the camera to start to track the puck.

I've never needed to call this before so like i say i'm unsure what has affected this from `v0.40` to `v0.42`

**Comment by Guardiola31337** [comment]
@Danny-James 
> I've managed to come up with a workaround for the camera/tracking issue

Glad to hear 🎉 Thanks for the update.

> I've never needed to call this before so like i say i'm unsure what has affected this from `v0.40` to `v0.42`

I think this is related to https://github.com/mapbox/mapbox-navigation-android/pull/1995 (introduced in `v0.41.0`). Could you share your setup so we can 👀 if this is a regression? As mentioned above, a complete example reproducing the issue would be 💯

Thanks for reporting 🙇

**Comment by Danny-James** [comment]
@Guardiola31337, see below it's a simple MapMatching Setup;

```java
@SuppressLint("MissingPermission")
@Override
protected void onCreate(@Nullable Bundle savedInstanceState) {
	
	setTheme(R.style.Theme_AppCompat_NoActionBar);
	super.onCreate(savedInstanceState);
	Fabric.with(this, new Crashlytics());

	// Mapbox access token is configured here. This needs to be called either in your application
	// object or in the same activity which contains the mapview.
	Log.d(TAG, "Mapbox.getInstace();");
	Mapbox.getInstance(getApplicationContext(), getString(R.string.mapbox_access_token));

	// Retrieve Coordinates From File.
	getCoordinatesData();

	setContentView(R.layout.activity_navigation);

	navigationView = findViewById(R.id.navigationView);
	navigationView.onCreate(savedInstanceState);
	navigationView.initialize(this);
}

@Override
public void onNavigationReady(boolean isRunning) {
	//if (!isRunning) {

	// Hide Feedback Button And Summary Sheet.
	navigationView.findViewById(R.id.feedbackFab).setVisibility(View.GONE);
	navigationView.findViewById(R.id.summaryBottomSheet).setVisibility(View.GONE);

	NavigationMapboxMap navigationMapboxMap = navigationView.retrieveNavigationMapboxMap();
	mapboxMap = navigationMapboxMap.retrieveMap();
	navigationMapboxMap.clearMarkers();

	// Set Default Starting Location
	mapboxMap.setCameraPosition(new CameraPosition.Builder()
			.target(new LatLng(53.481215, -2.244063)) // Default to Manchester UK.
			.zoom(6)
			.build());

	// Display Prompt To User To Check If Starting Partway Through Route.
	if (!isRunning) {
		showStartOfJourneyDialog();
	}
	// }else{
	//     Log.d(TAG, "Navigation Is Already Running - isRunning = TRUE");
	// }

}

private void fetchRoute(boolean showProgress, boolean routeFromCurrentLocation) {
	Log.d(TAG, "Start fetchRoute();");
	MapboxMapMatching.Builder mapMatchingBuilder = MapboxMapMatching.builder()
			.post()
			.accessToken(Mapbox.getAccessToken())
			.coordinates(subList)
			.steps(true)
			.tidy(true) // 1.4.0-beta -- Enabled to address issue with mini-roundabouts
			.voiceInstructions(true)
			.bannerInstructions(true)
			.language(Locale.getDefault().getLanguage())
			.voiceUnits(DirectionsCriteria.IMPERIAL)
			.waypoints(0, (subList.size() - 1))
			.profile(DirectionsCriteria.PROFILE_DRIVING);

		  if (points.size() > 1) {
			Log.d(TAG, "Start Mapmatching Request");
			mapMatchingBuilder.build().enqueueCall(new Callback<MapMatchingResponse>() {
				@Override
				public void onResponse(Call<MapMatchingResponse> call, Response<MapMatchingResponse> response) {
					Log.d(TAG, "Mapmatching Response Received");
					if (response.isSuccessful()) {
						Log.d(TAG, response.message());
						if (response.body().matchings().size() > 0) {

							Log.d(TAG, "MapMatching Completed Successfully.");								
							startNavigation(response.body().matchings().get(0).toDirectionRoute());

						} else {
							showUnableToRouteDialog("Unable to plot route. \n\n Issue with MapMatching Request.");
							Log.e(TAG, "No Matching Available");
						}
					} else {
						Log.e(TAG, "MapMatching Request Not Successful.");
						showUnableToRouteDialog("Unable to plot route. \n\n Unsuccessful Response. ");
						Log.e(TAG, response.message());
					}
				}

				@Override
				public void onFailure(Call<MapMatchingResponse> call, Throwable t) {
					showUnableToRouteDialog("Unable to plot route. \n\n" + t.getMessage());
					//navigationView.stopNavigation();
					Log.e(TAG, t.getMessage());
				}
			});
		}
	} else {
		Log.e(TAG, "No Coordinates Passed Into Navigation.");
		showUnableToRouteDialog("Unable to plot route.");
	}
}


@SuppressLint("RestrictedApi")
private void startNavigation(DirectionsRoute directionsRoute) {
	Log.d(TAG, "StartNavigation");
	NavigationViewOptions navigationViewOptions = setupOptions(directionsRoute);
	try {
		// Stop Navigation If Already Running.
		navigationView.stopNavigation();
	} catch (Exception ignored) {
	}

	// Start Navigation
	navigationView.startNavigation(navigationViewOptions);
	
	// Hide Feedback FAB
	navigationView.findViewById(R.id.feedbackFab).setVisibility(View.GONE);
	
	// Show Custom Exit FAB
	FloatingActionButton exitFab = findViewById(R.id.exitFab);
	exitFab.setVisibility(View.VISIBLE);

	// Reset navigationView Camera Position.
	navigationView.resetCameraPosition();

	// Hide Navigation Line And Show Custom Plotted Polyline.
	navigationView.retrieveNavigationMapboxMap().removeRoute();
	navigationView.retrieveNavigationMapboxMap().clearMarkers();
	navigationView.retrieveNavigationMapboxMap().addDestinationMarker(destination);

}

private NavigationViewOptions setupOptions(DirectionsRoute directionsRoute) {
	NavigationViewOptions.Builder options = NavigationViewOptions.builder();
	options.directionsRoute(directionsRoute)
			.navigationListener(this)
			.progressChangeListener(this)
			.routeListener(this)
			.shouldSimulateRoute(shouldSimulateRoute())
			.speechAnnouncementListener(this)
			.bannerInstructionsListener(this);

	return options.build();
}
```

**Comment by Danny-James** [comment]
@Guardiola31337 any update on this?

**Comment by Guardiola31337** [comment]
@abhishek1508 could you check if this would be an issue in `1.0`?

**Comment by abhishek1508** [comment]
Let's verify this.
cc @JunDai @cafesilencio @langsmith

**Comment by cafesilencio** [comment]
I tested with our basic NavigationView example activity for the soon to be released **1.0 SDK**. The source is available [here](https://github.com/mapbox/mapbox-navigation-android/blob/master/examples/src/main/java/com/mapbox/navigation/examples/ui/NavigationViewActivity.kt).

For point 1 in the initial issue I see the re-center button becoming visible when the map is dragged and tapping the button repositions the camera to follow the puck.

For point 2, there have been a number of improvements in the map matching since this issue was reported. When testing with the example I see the puck following the route.

**Comment by cafesilencio** [comment]
Please be on the lookout for our first public preview of UI SDK v1.0 coming soon. Feel free to create a new issue if you feel it still exists.
